examples/wakaama/dtlsconnection.c

```diff
@@ -0,0 +1,570 @@
+   }
+   /* allocate IPv6 header */
+   ip = gnrc_ipv6_hdr_build(udp, NULL, &addr);
+   if (ip == NULL) {
+      puts("Error: unable to allocate IPv6 header");
+      gnrc_pktbuf_release(udp);
+      return -1;
+   }
+   /* add netif header, if interface was given */
+   if (iface > 0) {
+      gnrc_pktsnip_t *netif = gnrc_netif_hdr_build(NULL, 0, NULL, 0);
+
+      ((gnrc_netif_hdr_t *)netif->data)->if_pid = (kernel_pid_t)iface;
+      LL_PREPEND(ip, netif);
+   }
+   /* send packet */
+
+   DEBUG("Sending stuff via gnrc_sending(%s, %d)\n", addr_str, rem_port);
+
+   /*
+    * WARNING: Too fast and the nodes dies in middle of retransmissions.
+    *         This issue appears in the FIT-Lab (m3 motes).
+    */
+   xtimer_usleep(500000);
+
+   /* Probably this part will be removed.  **/
+   if (!gnrc_netapi_dispatch_send(GNRC_NETTYPE_UDP, GNRC_NETREG_DEMUX_CTX_ALL,
+                                  ip)) {
+      puts("Error: unable to locate UDP thread");
+      gnrc_pktbuf_release(ip);
+      return -1;
+   }
+
+   return 1;
+}
+
+/**************************   RIOT GNRC Ends  ********************************/
+
+static dtls_handler_t cb = {
+    .write = send_to_peer,
+    .read = read_from_peer,
+    .event = NULL,
+    //#ifdef DTLS_PSK
+    // FIXME: other key stuff
+    .get_psk_info = get_psk_info,
+    //#endif /* DTLS_PSK */
+    //#ifdef DTLS_ECC
+    //  .get_ecdsa_key = get_ecdsa_key,
+    //  .verify_ecdsa_key = verify_ecdsa_key
+    //#endif /* DTLS_ECC */
+};
+
+dtls_context_t *get_dtls_context(dtls_connection_t *connList) {
+   if (dtlsContext == NULL) {
+      dtls_init();
+      dtlsContext = dtls_new_context(connList);
+      if (dtlsContext == NULL)
+         fprintf(stderr, "Failed to create the DTLS context\r\n");
+      dtls_set_handler(dtlsContext, &cb);
+   } else {
+      dtlsContext->app = connList;
+   }
+   return dtlsContext;
+}
+
+dtls_connection_t *connection_find(dtls_connection_t *connList,
+                                   const ipv6_addr_t *addr,
+                                   unsigned short rem_port) {
+   dtls_connection_t *connP;
+
+   connP = connList;
+   while (connP != NULL) {
+      if (connP->dtlsSession->port == rem_port &&
+          memcmp(&connP->addr, addr, sizeof(ipv6_addr_t)) == 0) {
+         return connP;
+      }
+
+      connP = connP->next;
+   }
+
+   return connP;
+}
+
+dtls_connection_t *connection_new_incoming(dtls_connection_t *connList,
+                                           const ipv6_addr_t *addr,
+                                           unsigned short rem_port) {
+   dtls_connection_t *connP;
+
+   connP = (dtls_connection_t *)malloc(sizeof(dtls_connection_t));
+   if (connP != NULL) {
+      memset(connP, 0, sizeof(dtls_connection_t));
+      memcpy(&(connP->addr), addr, sizeof(ipv6_addr_t));
+      connP->addrLen = sizeof(ipv6_addr_t);
+      connP->next = connList;
+      connP->noSec = false;
+
+      connP->dtlsSession = (session_t *)malloc(sizeof(session_t));
```

Comment:
Eventually you need to remove the malloc/free calls. Maybe PR #7651 can be used here.

Reply:
The Wakaama package itself uses malloc/free as well, so it would probably be a relatively big undertaking to move all that to memarray. Not sure that falls under the scope of this pull request, especially since memarray isn't merged yet.

Reply:
Sadly, if you stay with malloc/free, there will be a lot of problems with many boards.  Because  is not that well supported on many RTOS (including RIOT). You can use malloc() the first time, but [free() is not going to work](http://riot-os.org/api/group__oneway__malloc.html#ga7ac38fce3243a7dcf448301ee9ffd392). 

Also, if you are not sure of having enough memory, you will have a nightmare catching the overflows.

Reply:
Alright, that's a very good argument to do this sooner rather than later. I just rebased the changes from #7651 into my branch and started moving things over. I'll have to see what to do about variable sized buffer allocations and such.

I'll also have to look into how to best apply this to the Wakaama package.

Reply:
I just replaced all the malloc/free stuff with memarray. Though I'm not sure how hacky/evil the thing is I did for lwm2m_malloc() and lwm2m_free() in platform.c for variable sized allocations.